Documentation/design-docs/dllmap.md

```diff
@@ -0,0 +1,246 @@
+    
+**AssemblyLoadContext [unmanaged code]**
+-   Defines `LoadLibraryModuleViaCallbackand` and exposes an API:
+	```c++
+	IntPtr  LoadLibraryModuleViaCallback(string libraryName)
+	IntPtr  GetEntrypointViaCallback(string entrypointName, HMOD hmod)	
+	```
+
+**Dotnet.corefx.labs.dll [managed code]**
+-   Implements LoadNativeDllViaDllMap and GetEntrypointViaDllMap
+	```c++
+	IntPtr  LoadNativeDllViaDllMap(string libraryName)
+	{
+		if (libraryName in cachedResults)
+			return cachedResults[libraryName];
+		if (!mapStructure)
+			mapStructure = ReadAndParseXML();
+		targetLibraryName = mapStructure.GetLibrary(libraryName);
+		hmod = LoadLibrary(targetLibraryName);
+		AddToCache(libraryName, hmod);
+		return hmod;
+	}
+	```
+	```c++
+	IntPtr  GetEntrypointViaDllMap(string entrypointName, HMOD hmod)
+	{
+		targetEntrypointName = mapStructure.GetEntrypoint(entrypointName);
+		pvTarget = GetProcAddress(targetEntrypointName, hmod);
+		return pvTarget;
+	}
+	```
+
+## Testing
+
+To verify if Dllmap’s behavior is correct, cross-platform tests will be run: [Mono tests](https://github.com/mono/mono/tree/0bcbe39b148bb498742fc68416f8293ccd350fb6/mcs/tools/mono-shlib-cop) and custom tests created on purpose of this feature.
+Custom tests must contain calls to native libraries. It doesn’t really matter what functions from which libraries get called, because the feature is about mapping things independently from method and library exact names. There are multiple dimensions that must be included in tests:
+
+**OS**. When it comes to the OS, there are three main target platforms for this feature: Windows, Linux and OS X. All the Linux based systems are so architecturally similar, so they are named just “Linux” in this document. Ubuntu 16.04LTS will be used for testing.
+Test cases:
+
+|Develop (initial dll)  |Run (target dll)|
+|--|--|
+|Windows |Windows  |
+|Windows |Linux |
+|Windows |OS X |
+|Linux | Windows|
+|Linux |OS X |
+|Linux |OS X |
+|OS X |Windows|
+|OS X |Linux |
+|OS X |OS X
+
+**System's architecture** – test cases:
+
+|Develop |Run |
+|--|--|
+|32-bit |32-bit|
+|32-bit|64-bit|
+|64-bit|32-bit |
+|64-bit| 32-bit|
+
+**Path mapping** – test cases:
+
+|Initial dll path |Target dll path|
+|--|--|
+|absolute |absolute|
+|absolute|relative|
+|relative|absolute |
+|relative| relative|
+ 
+**Dll naming** – test cases:
+
+|Initial dll name |Target dll nanme|
+|--|--
+|with extension (foo.dll)|with extension|
+|with extension|no extension|
+|no extension (foo)|with extension|
+|no extension| no extension|
+
+All the above test cases will be covered.
+
+### Related discussion
+[Handling p/invokes for different platforms and discussions about dllmap](https://github.com/dotnet/coreclr/issues/930)
+[Add NativeLibrary class PR](https://github.com/dotnet/coreclr/pull/16409/commits/7ece113b5f58111ee934d923e1ea213ba50f4224)
```

Comment:
Most recent related discussion is in https://github.com/dotnet/coreclr/issues/19112